sdk/core/azure-core/azure/core/pipeline/_tools_async.py

```diff
@@ -23,13 +23,27 @@
 # IN THE SOFTWARE.
 #
 # --------------------------------------------------------------------------
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Callable, TypeVar, Awaitable, Union, overload
+from typing_extensions import ParamSpec
 
 if TYPE_CHECKING:
     from ..rest import AsyncHttpResponse as RestAsyncHttpResponse
 
+P = ParamSpec("P")
+T = TypeVar("T")
 
-async def await_result(func, *args, **kwargs):
+
+@overload
+async def await_result(func: Callable[P, Awaitable[T]], *args: P.args, **kwargs: P.kwargs) -> T:
+    ...
+
+
+@overload
+async def await_result(func: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
```

Comment:
Are the overloads necessary? The type of `func` has no impact on the rest of args/output. So wouldn't the `Union` definition in the base method signature suffice?

Reply:
I was not able to make it work without the overload
```
Argument 1 to "await_result" has incompatible type "Callable[[PipelineRequest[HttpRequest]], Coroutine[Any, Any, None]]"; expected "Callable[[PipelineRequest[HttpRequest]], Awaitable[<nothing>]]"
```

cc @kristapratico

Reply:
Feels like this is an instance of https://github.com/python/mypy/issues/14669

The overload makes mypy happy, I'll be pragmatic :)